File: trump_game/code4.py
import random

class Card:
      def __init__(self, value, suit, intValue) :
            self.value = value
            self.suit = suit
            self.intValue = intValue

      def getCardString(self):
            return self.suit + self.value + "(" + str(self.intValue) + ")"

class Deck:
      def __init__(self):
            self.deck = self.generateDeck()

      def printDeck(self):
            print("Displaying cards...")
            for i in self.deck:
                  print(i.getCardString())

      def shuffleDeck(self):
            deckSize = len(self.deck)
            for i in reversed(range(0, deckSize)):
                  j = random.randint(0,i)
                  temp = self.deck[i]
                  self.deck[i] = self.deck[j]
                  self.deck[j] = temp

      # デッキからカードを1枚引く関数を作成しましょう。配列の末尾を取得するとともに、削除する処理を入れます。
      def draw(self):
            return self.deck.pop()

      @staticmethod
      def generateDeck():
            newDeck = []
            values = ["A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"]
            suits = ["♣", "♦", "♥", "♠"]

            for suit in suits:
                  for i, value in enumerate(values):
                        newDeck.append(Card(value, suit, i))

            return newDeck


# ディーラーの動きをまとめたクラスDealerを作成します。まずはゲームスタート時にテーブルの状態（各プレイヤーの手札、ゲームの種類、シャッフルされたデッキ）を返す関数を作成しましょう。
# その際、クラスDeck内にカードを1枚引く関数と、クラスDealer内にゲームの種類によって引く手札の数を返す関数が必要になります。先に作成してから、テーブルの状態を返す関数を作成してください。

# ここから記入してください
# インスタンスを作成する必要がないので、全てstaticメソッドにしましょう
class Dealer:

      # テーブルの状態を表す関数です。人数とゲームの種類が与えられ、デッキのシャッフルと各プレイヤーの手札を作成する処理を行い、テーブルの状態を返します
      @staticmethod
      def initialTable(amoutOfPeople ,gameMode):
            # 連想配列でテーブルの状態を作成します。
            table = {
                  "players": [],
                  "gameMode": gameMode,
                  "deck": Deck()
            }

            # デッキをシャッフルします
            table["deck"].shuffleDeck()

            # デッキからカードを引いていき、各プレイヤーの手札をtableにセットします。
            for person in range(0, amoutOfPeople):
                  personCards = []
                  for cards in range(0, Dealer.initialCards(gameMode)):
                        personCards.append(table["deck"].draw())
                  table["players"].append(personCards)

            return table


      # ゲームの種類が与えられるので、種類によって何枚カードを引くか、数字を返す関数を作成しましょう。
      @staticmethod
      def initialCards(gameMode):
            if gameMode == "21":
                  return 2
            if gameMode == "porker":
                  return 5

      # テーブルの中身（各プレイヤーの手札、ゲームの種類、シャッフルされたデッキ）を表示する関数です。ディーラークラス作成後、これを使って正常な処理ができているか確認してください
      @staticmethod
      def printTable(table):
            print("This table: ")
            print("gameMode: " + table["gameMode"])
            print("deck: ")
            table["deck"].printDeck()
            for i, player in enumerate(table["players"]):
                  print(str(i + 1) + "player's cards: ")
                  for cards in player:
                        print(cards.getCardString())

# テーブルを生成し、どのような状態か出力します
table1 = Dealer.initialTable(3, "21")
Dealer.printTable(table1)

File: trump_game_js/code4.js
class Card {
      constructor(suit, value, intValue) {
            this.suit = suit;
            this.value = value;
            this.intValue = intValue
      }

      getCardString() {
            return this.suit + this.value + "(" + this.intValue + ")";
      }
}

class Deck {
      constructor() {
            this.deck = Deck.generateDeck();
      }

      shuffleDeck() {
            for (let i = 0; i < this.deck.length; i++) {
                  let j = Math.floor(Math.random() * (i + 1));
                  let temp = this.deck[i];
                  this.deck[i] = this.deck[j];
                  this.deck[j] = temp;
            }
      }

      printDeck() {
            console.log("Displaying cards...")
            for (let i = 0; i < this.deck.length; i++) {
                  console.log(this.deck[i].getCardString());
            }
      }

      // デッキからカードを1枚引く関数を作成しましょう。配列の末尾を取得するとともに、削除する処理を入れます。
      draw() {
            return this.deck.pop()
      }

      static generateDeck() {
            let newDeck = [];
            const suits = ["♣", "♦", "♥", "♠"];
            const values = ["A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K"]

            for (let i = 0; i < suits.length; i++) {
                  for (let j = 0; j < values.length; j++) {
                        newDeck.push(new Card(suits[i], values[j], j));
                  }
            }

            return newDeck;
      }
}

// ディーラーの動きをまとめたクラスDealerを作成します。まずはゲームスタート時にテーブルの状態（各プレイヤーの手札、ゲームの種類、シャッフルされたデッキ）を返す関数を作成しましょう。
// その際、クラスDeck内にカードを1枚引く関数と、クラスDealer内にゲームの種類によって引く手札の数を返す関数が必要になります。先に作成してから、テーブルの状態を返す関数を作成してください。

// ここから記入してください
// インスタンスを作成する必要がないので、全てstaticメソッドにしましょう
class Dealer {

      // テーブルの状態を表す関数です。人数とゲームの種類が与えられ、デッキのシャッフルと各プレイヤーの手札を作成する処理を行い、テーブルの状態を返します
      static startGame(amountOfPlayers, gameMode) {
            // 連想配列でテーブルの状態を作成します。
            let table = {
                  "players":[],
                  "gameMode": gameMode,
                  "deck": new Deck()
            }

            // デッキをシャッフルします
            table["deck"].shuffleDeck();

            // デッキからカードを引いていき、各プレイヤーの手札をtableにセットします。
            for (let i = 0; i < amountOfPlayers; i++) {
                  let playerCard = [];
                  for (let j = 0; j < Dealer.initialCards(gameMode); j++) {
                        playerCard.push(table["deck"].draw());
                  }
                  table["players"].push(playerCard);
            }

            return table;
      }

      // ゲームの種類が与えられるので、種類によって何枚カードを引くか、数字を返す関数を作成しましょう。
      static initialCards(gameMode) {
            if (gameMode == "21") { return 2;}
            if (gameMode == "porker") { return 5;}
      }

      // テーブルの状態を表す関数を用意しました。これを使って、テーブルの状態を確認しましょう
      static printTableInformation(table) {
            console.log("Amount of players: " + table["players"].length + "... Game mode: " + table["gameMode"] + ". At this table: ");

            for (let i = 0; i < table["players"].length; i++) {
                  console.log("Player " + (i + 1) + " hand is: ");
                  for(let j = 0; j < table["players"][i].length; j++) {
                        console.log(table["players"][i][j].getCardString());
                  }
            }
      }

      
}

// 新しくデッキを生成します
let table1 = Dealer.startGame(3, "21");
// 用意した関数を使って、テーブルの状態を見ましょう
Dealer.printTableInformation(table1);